File: javascript/project.js
// Copyright 2014 Tom SF Haines

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.



// Functions for adding and removing roles...
function add_role()
{
 // Callback function...
  var callback = function(response)
  {
   if (response)
   {
    window.location.reload(); 
   }
   else
   {
    alert('Error adding role'); 
   }
  }
  
 // Extract needed parameters...
  var role = $('#add_role input').val();
  var user = $('#add_role select').val();
 
 // Do the call...
  $.getJSON('/add/role', {'role':role, 'user':user}, callback);
}

function remove_role()
{
 // Callback function...
  var callback = function(response)
  {
   if (response)
   {
    window.location.reload(); 
   }
   else
   {
    alert('Error removing role'); 
   }
  }
  
 // Figure out the ident...
  var ident = $(this).parent().parent().attr('id');
  
 // Send the request...
  $.getJSON('/remove/role', {'ident':ident}, callback);
}


// Functions to move a role...
function role_up()
{
 // Callback function...
  var callback = function(response)
  {
   if (response)
   {
    window.location.reload(); 
   }
   else
   {
    alert('Error moving a role up'); 
   }
  }
 
 // Figure out the ident...
  var ident = $(this).parent().parent().attr('id');
  
 // Send the request...
  $.getJSON('/store/role/'+ident, {'key':'up'}, callback);
}

function role_down()
{
 // Callback function...
  var callback = function(response)
  {
   if (response)
   {
    window.location.reload(); 
   }
   else
   {
    alert('Error moving a role down'); 
   }
  }
 
 // Figure out the ident...
  var ident = $(this).parent().parent().attr('id');
  
 // Send the request...
  $.getJSON('/store/role/'+ident, {'key':'down'}, callback);
}



// Functions for adding adn removing assets...
function add_ext_asset()
{
 // Callback function...
  var callback = function(response)
  {
   if (response)
   {
    window.location.reload(); 
   }
   else
   {
    alert('Error adding external asset'); 
   }
  }
  
 // Extract needed parameters...
  var description = $('#add_ext_asset .asset_description input').val();
  var license = $('#add_ext_asset .asset_license input').val();
  var origin = $('#add_ext_asset .asset_origin input').val();
 
 // Do the call...
  $.getJSON('/add/ext_asset', {'description' : description, 'license' : license, 'origin' : origin}, callback);
}

function remove_ext_asset()
{
 // Callback function...
  var callback = function(response)
  {
   if (response)
   {
    window.location.reload(); 
   }
   else
   {
    alert('Error removing external asset'); 
   }
  }
  
 // Figure out the ident...
  var ident = $(this).parent().parent().attr('id');
  
 // Send the request...
  $.getJSON('/remove/ext_asset', {'ident':ident}, callback);
}



// Setup all the handlers...
$(function() {
 // The basic ones...
  tie_input('#title', 'project', 'title');
  tie_input('#description', 'project', 'description');
  tie_input('#width', 'project', 'width');
  tie_input('#height', 'project', 'height');
  tie_input('#fps', 'project', 'fps');
  tie_input('#license', 'project', 'license');
 
 // The role stuff...
  $('#add_role button').click(add_role);
  $('.role_row button').click(remove_role);
  
  $('.role_row .up').click(role_up);
  $('.role_row .down').click(role_down);
  
  $('.role_row').each(function()
  {
   var ident = $(this).attr('id');
   
   tie_input('#' + ident + ' .role_role', 'role/' + ident, 'role');
   tie_input('#' + ident + ' .role_user', 'role/' + ident, 'user');
  });
  
 // The external asset stuff...
  $('#add_ext_asset button').click(add_ext_asset);
  $('.ext_asset button').click(remove_ext_asset);
  
  $('.ext_asset').each(function()
  {
   var ident = $(this).attr('id');
   
   tie_input('#' + ident + ' .asset_description', 'ext_asset/' + ident, 'description');
   tie_input('#' + ident + ' .asset_license', 'ext_asset/' + ident, 'license');
   tie_input('#' + ident + ' .asset_origin', 'ext_asset/' + ident, 'origin');
  });
});

File: bin/project.py
# Copyright 2014 Tom SF Haines

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import xml.sax.saxutils as saxutils

from .login import app as login



def app(rfam, response):
  # If the user is not logged in show them the login page...
  if response.user==None or response.project==None:
    login(rfam, response)
    return
  
  # Access the project...
  db = rfam.proj(response.project)
  
  if 'project.json' in db:
    p = db['project.json'].read()
  else:
    p = None
    
  if p==None:
    p = {'title' : rfam.getLanguage(response.user)['default_title'], 'description' : '', 'roles' : {}, 'ext_assets' : {}, 'license' : rfam.getLanguage(response.user)['default_license'], 'default' : rfam.getDefault()}
    db.get_root().new('project.json', p)
    
  # Prepare the defaults part of the page...
  choice_default = rfam.defaultChoice(p['default'])
  
  # Prepare the add roles part of the page...
  choice_people = rfam.userChoice(response.project)
  
  # Prepare the roles part of the page...
  roles = []
  
  for key, value in sorted(p['roles'].items(), key = lambda p: p[1]['order']):
    user_choice = rfam.userChoice(response.project, value['user'])
    payload = {'ident' : key, 'role' : saxutils.escape(value['role']), 'user' : user_choice}
    roles.append(rfam.template('role', payload, response))
  
  roles = '\n'.join(roles)
  
  # Prepare the external assets part of the page...
  assets = []
  
  for key, value in sorted(p['ext_assets'].items(), key=lambda p: p[1]['description']):
    payload = {'ident' : key, 'description' : saxutils.escape(value['description']), 'license' : saxutils.escape(value['license']), 'origin' : saxutils.escape(value['origin'])}
    assets.append(rfam.template('asset.ext', payload, response))
  
  assets = '\n'.join(assets)
  
  # Prepare the automatic credits button...
  credits_button = rfam.template('button.credits', {}, response)
  
  # Show the project page...
  head = '<link rel="stylesheet" href="/stylesheets/project.css"/>\n<script src="/javascript/project.js"></script>'
  payload = {'title' : rfam.getLanguage(response.user)['project'], 'head' : head, 'project_title' : saxutils.escape(p['title']), 'project_description' : saxutils.escape(p['description']), 'project_license' : saxutils.escape(p['license']), 'roles' : roles, 'choice_people' : choice_people, 'ext_assets' : assets, 'choice_default' : choice_default, 'header_extra' : credits_button}
  html = rfam.template('project', payload, response)
  
  response.append(html)
  response.setHTML()
